L1Trigger/Phase2L1ParticleFlow/plugins/DeregionizerProducer.cc

```diff
@@ -0,0 +1,168 @@
+#include <unordered_map>
+
+#include "FWCore/Framework/interface/stream/EDProducer.h"
+#include "FWCore/Framework/interface/Event.h"
+#include "FWCore/Utilities/interface/InputTag.h"
+#include "FWCore/ParameterSet/interface/ParameterSet.h"
+
+#include "DataFormats/L1TParticleFlow/interface/PFCandidate.h"
+
+#include "L1Trigger/Phase2L1ParticleFlow/src/newfirmware/deregionizer/deregionizer_input.h"
+#include "L1Trigger/Phase2L1ParticleFlow/src/newfirmware/deregionizer/deregionizer_ref.h"
+
+
+class DeregionizerProducer : public edm::stream::EDProducer<> {
+  public:
+    explicit DeregionizerProducer(const edm::ParameterSet &);
+    ~DeregionizerProducer() override;
+
+  private:
+    edm::ParameterSet config_;
+    edm::EDGetTokenT<l1t::PFCandidateRegionalOutput> token_;
+    l1ct::DeregionizerEmulator emulator_;
+    bool debug_;
+
+    std::unordered_map<const l1t::PFCandidate *, l1t::PFClusterRef> clusterRefMap_;
+    std::unordered_map<const l1t::PFCandidate *, l1t::PFTrackRef> trackRefMap_;
+    std::unordered_map<const l1t::PFCandidate *, l1t::PFCandidate::MuonRef> muonRefMap_;
+
+    void produce(edm::Event &, const edm::EventSetup &) override;
+    void hwToEdm_(const std::vector<l1ct::PuppiObjEmu> &hwOut, std::vector<l1t::PFCandidate> &edmOut) const;
+    void setRefs_(l1t::PFCandidate &pf, const l1ct::PuppiObjEmu &p) const;
+};
+
+DeregionizerProducer::DeregionizerProducer(const edm::ParameterSet &iConfig)
+  : config_(iConfig),
+    token_(consumes<l1t::PFCandidateRegionalOutput>(iConfig.getParameter<edm::InputTag>("RegionalPuppiCands"))),
+    emulator_(iConfig),
+    debug_(iConfig.getUntrackedParameter<bool>("debug", 0)) {
+  produces<l1t::PFCandidateCollection>("Puppi");
+  produces<l1t::PFCandidateCollection>("TruncatedPuppi");
+}
+
+DeregionizerProducer::~DeregionizerProducer() {}
+
+void DeregionizerProducer::produce(edm::Event &iEvent, const edm::EventSetup &iSetup) {
+  clusterRefMap_.clear();
+  trackRefMap_.clear();
+  muonRefMap_.clear();
+
+  auto deregColl = std::make_unique<l1t::PFCandidateCollection>();
+  auto truncColl = std::make_unique<l1t::PFCandidateCollection>();
+
+  edm::Handle<l1t::PFCandidateRegionalOutput> src;
+
+  iEvent.getByToken(token_, src);
+
+  std::vector<float> regionEtas, regionPhis;
+  std::vector<l1ct::OutputRegion> outputRegions;
+  std::vector<l1ct::PuppiObjEmu> hwOut;
+  std::vector<l1t::PFCandidate> edmOut;
+  std::vector<l1ct::PuppiObjEmu> hwTruncOut;
+  std::vector<l1t::PFCandidate> edmTruncOut;
+
+  if (debug_) std::cout << "\nRegional Puppi Candidates" << "\n";
```

Comment:
Here since this is a CMSSW module, you should replace `std::cout` with `edm::LogPrint("DeregionizerProducer")`, (and remove the trailing `"\n"` since the logger will add it automatically).